fdbserver/ClusterController.actor.cpp

```diff
@@ -337,19 +337,56 @@ class ClusterControllerData {
 
 		logServerSet = Reference<LocalitySet>(new LocalityMap<WorkerDetails>());
 		logServerMap = (LocalityMap<WorkerDetails>*)logServerSet.getPtr();
-		for (auto& it : id_worker) {
-			if (std::find(exclusionWorkerIds.begin(), exclusionWorkerIds.end(), it.second.details.interf.id()) ==
+
+		// Populate `unavailableLocals` and log the reason why the worker is considered as unavailable.
+		auto logWorkerUnavailable = [this, &unavailableLocals](const std::string& reason,
+		                                                       const WorkerDetails& details,
+		                                                       ProcessClass::Fitness fitness) {
+			unavailableLocals.push_back(details.interf.locality);
+
+			// Note that the recruitment happens only during initial database creation and recovery. So these trace
+			// events should be sparse.
+			// TODO(zhewu): Add targeting dcids.
+			TraceEvent("GetTLogTeamWorkerUnavailable", id)
+			    .detail("Reason", reason)
+			    .detail("WorkerID", details.interf.id())
+			    .detail("WorkerDC", details.interf.locality.dcId())
+			    .detail("Address", details.interf.addresses().toString())
+			    .detail("Fitness", fitness);
+		};
+
+		// Go through all the workers to list all the workers that can be recruited.
+		for (const auto& [worker_process_id, worker_info] : id_worker) {
+			const auto& worker_details = worker_info.details;
+			if (std::find(exclusionWorkerIds.begin(), exclusionWorkerIds.end(), worker_details.interf.id()) !=
 			    exclusionWorkerIds.end()) {
-				auto fitness = it.second.details.processClass.machineClassFitness(ProcessClass::TLog);
-				if (workerAvailable(it.second, checkStable) &&
-				    !conf.isExcludedServer(it.second.details.interf.addresses()) &&
-				    fitness != ProcessClass::NeverAssign &&
-				    (!dcIds.size() || dcIds.count(it.second.details.interf.locality.dcId()))) {
-					fitness_workers[std::make_pair(fitness, it.second.details.degraded)].push_back(it.second.details);
-				} else {
-					unavailableLocals.push_back(it.second.details.interf.locality);
-				}
+				logWorkerUnavailable("Worker is excluded.", worker_details, ProcessClass::UnsetFit);
+				continue;
+			}
+
+			auto fitness = worker_details.processClass.machineClassFitness(ProcessClass::TLog);
+			if (!workerAvailable(worker_info, checkStable)) {
+				logWorkerUnavailable("Worker is not available.", worker_details, fitness);
+				continue;
+			}
+
+			if (conf.isExcludedServer(worker_details.interf.addresses())) {
+				logWorkerUnavailable("Worker's server is excluded from the cluster.", worker_details, fitness);
```

Comment:
do not use special character in the trace logs. It may break the Joshua parser and cause correctness failure.

Reply:
Worker's --> Worker
remove the . in the end

Reply:
If there are special characters that break the trace logs, we might want to update the trace log formatter to properly escape or remove them.

Reply:
In particular, I think some of our trace strings come from sources we don't control, so this becomes a difficult thing to enforce.

Reply:
Also for what it's worth, the `.` character at least is currently in use in other trace events.

Reply:
Unfortunately, I don't have a complete list.
I used to add "?" into trace event. That breaks the test painfully. :(

Reply:
It looks like `?` is legal in XML, so I guess it's ok that we don't escape it. It's not legal for our type names or for our field names, which I think is quasi-enforced by using asserts in simulation.

If Joshua has a problem with a `?` in the field values, that's probably a bug in Joshua. If you had used `?` in a field or type name, then it's probably not crazy that Joshua didn't like it and maybe we could try to find some better way to validate that these names are ok.

Reply:
Done.

It's probably worth addressing this in the TraceEvent (adding some sanity check for string format), given how frequent the TraceEvent is used.